### ISSUE TITLE
Clarification about stream connectors

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#378]
**Reported by svorkoetter on 9 Jun 2010 08:59 UTC**
Section 15.1 of the 3.1 spec says:

"If at least one variable in a connector has the stream prefix, the connector is called “stream connector” and the corresponding variable is called “stream variable”."

So clearly, the following is a stream connector:
```text/x-modelica
connector MyStreamConnector
  flow Real x;
  Real y;
  stream Real z;
end MyStreamConnector;
```
But, is this a stream connector?:
```text/x-modelica
connector AnotherConnector
  MyStreamConnector s;
  Real t;
end AnotherConnector;
```
After all, once AnotherConnector has been flattened, it is really,
```text/x-modelica
connect AnotherConnector
  flow Real s.x;
  Real s.y;
  stream Real s.z;
  Real t;
end AnotherConnector;
```
which satisfies the definition from section 15.1.

In other words, does the presence of a stream connector within a second connector make the second connector a stream connector? 

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/378

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 9 Jun 2010 09:02 UTC**
<sub>Only one comment from Thomas Eric:</sub>

For me, it was possible to create connectors using several stream connectors. Especially, I created one which is as follows:
```text/x-modelica
connector MyConnector
	FluidPort Water;
	FluidPort Air;
	Record PortData;
End MyConnector;
```
For me it was like creating tree independent connectors with only one link in the sketch. And typically Dymola seems to recognize the sub connectors and its interface allows connecting a part of it to make connection like : `connect(MyConnector.Water, DynamicPipe.Port_a)`
I find that it is a very interesting feature which mainly avoid creation at system level (vehicle) many useless links, by replacing them by aggregated connections where many things could be put (water, gas, electricity ...)

Therefore, I think that it should be discussed a little be more. It seems to work like that in Dymola. Am I really right ?

**Comment by modelica-trac-importer** [comment]
**Comment by fcasella on 9 Jun 2010 09:06 UTC**
Of course this kind of usage makes sense, and it should be explicitly allowed by the semantics of the connectors.

The idea is that if the connector contains structured entities, the matching between flow variables and strema variables should be performed within each of those entities. Though I'm not sure about the exact wording (are they really entities?)

**Comment by modelica-trac-importer** [comment]
**Comment by HansOlsson on 9 Jun 2010 09:33 UTC**
The connector containing two stream-connectors should be allowed, but cannot be a stream-connector, since a stream-connector must have exactly one variable with the flow prefix.

Thus to be consistent a connector containing a stream-connector is not a stream-connector, i.e. AnotherConnector above is not a stream connector.

_If_ it is important to allow some operation on AnotherConnector (as if it were a stream-connector) we have to introduce two different concepts:
* "Elementary" stream connector (the one containing exactly one flow).
* "Nested" stream connector: an "elementary" stream connector or a connector containing one or more "nested" stream-connector components.

I don't see where "nested" stream connector would be used in the specification, but might have missed some use-case. 


As for clarifying stream connectors the obvious other questions is if 'flow' and 'stream' must be at the same level or whether the following would be allowed:

```
connector A
  stream Real x;
end A;
connector B
  flow Real f;
end B;
connector MyPossibleStream
  A a1,a2;
  B b;
end MyPossibleStream;
```

When flattened this will satisfy the requirements, but neither A nor B are good connectors on their own.

I interpret the Definition of Stream connector to forbid MyPossibleStream (and AnotherConnector above) by requiring that the flow and stream variables are declared in the same connector. It might be possible to make it even more explicit that 'variable in a connector' never includes variables nested inside other variables.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 29 Mar 2016 15:27 UTC**
I am unsure if there is any unresolved issue left for these nested stream connectors, and would thus propose to close this as expired (but could see minor clarification that as in previous reply).

**Comment by modelica-trac-importer** [comment]
**Changelog removed by hansolsson on 29 Mar 2016 15:27 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 23 Jun 2016 12:59 UTC**
Housekeeping to make it easier to find the active tickets for the specification.

These tickets are all expired, i.e. old with no action - and not clear that any action is needed. In most cases they are enhancements/discussions without any clear result.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 26 Aug 2016 06:21 UTC**
House keeping.
